storage/src/key_value/agent/block_commit.rs

```diff
@@ -0,0 +1,291 @@
+            _ => (),
+        }
+
+        let mut sn_index = self.read_meta_u32(KEY_CURR_SN_INDEX, Some(0))?;
+        let mut cm_index = self.read_meta_u32(KEY_CURR_CM_INDEX, Some(0))?;
+        let mut memo_index = self.read_meta_u32(KEY_CURR_MEMO_INDEX, Some(0))?;
+
+        let mut transaction_cms = vec![];
+
+        for (index, transaction) in block.transactions.iter().enumerate() {
+            let cms = self.commit_transaction(&mut sn_index, &mut cm_index, &mut memo_index, transaction)?;
+            transaction_cms.extend(cms);
+
+            let transaction_location = TransactionLocation {
+                index: index as u32,
+                block_hash: block_hash.clone(),
+            };
+            let mut out = vec![];
+            transaction_location.write(&mut out)?;
+            self.inner
+                .store(KeyValueColumn::TransactionLookup, &transaction.id[..], &out[..])?;
+        }
+
+        self.write_meta_u32(KEY_CURR_SN_INDEX, sn_index)?;
+        self.write_meta_u32(KEY_CURR_CM_INDEX, cm_index)?;
+        self.write_meta_u32(KEY_CURR_MEMO_INDEX, memo_index)?;
+
+        let is_genesis = canon.block_height == 0 && canon.hash.is_empty();
+
+        let new_best_block_number = if is_genesis { 0 } else { canon.block_height as u32 + 1 };
+
+        self.write_meta_u32(KEY_BEST_BLOCK_NUMBER, new_best_block_number)?;
+
+        let block_num_serialized = &new_best_block_number.to_le_bytes()[..];
+
+        self.inner
+            .store(KeyValueColumn::BlockIndex, &block_hash[..], block_num_serialized)?;
+        self.inner
+            .store(KeyValueColumn::BlockIndex, block_num_serialized, &block_hash[..])?;
+
+        self.inner
+            .store(KeyValueColumn::DigestIndex, &ledger_digest[..], block_num_serialized)?;
+        self.inner
+            .store(KeyValueColumn::DigestIndex, block_num_serialized, &ledger_digest[..])?;
+
+        Ok(BlockStatus::Committed(new_best_block_number as usize))
+    }
+
+    pub(super) fn decommit_transaction(
+        &mut self,
+        sn_index: &mut u32,
+        cm_index: &mut u32,
+        memo_index: &mut u32,
+        transaction: &SerialTransaction,
+    ) -> Result<()> {
+        for serial in transaction.old_serial_numbers.iter() {
+            self.inner.delete(KeyValueColumn::SerialNumber, &serial[..])?;
+            *sn_index -= 1;
+        }
+
+        for commitment in transaction.new_commitments.iter() {
+            self.inner.delete(KeyValueColumn::Commitment, &commitment[..])?;
+            *cm_index -= 1;
+        }
+
+        self.inner
+            .delete(KeyValueColumn::TransactionLookup, &transaction.id[..])?;
+
+        self.inner.delete(KeyValueColumn::Memo, &transaction.memorandum[..])?;
+        *memo_index -= 1;
+
+        Ok(())
+    }
+
+    pub(super) fn decommit_blocks(&mut self, hash: &Digest) -> Result<Vec<SerialBlock>> {
+        match self.get_block_state(hash)? {
```

Comment:
`clippy` will probably complain about using `if let` for single-branch matches here